Toolbar BETA Toggle

mariusandra opened the pull request:
## Changes

This adds a toggle under /setup to enable/disable the new toolbar:

![image](https://user-images.githubusercontent.com/53387/84018799-bbce7800-a980-11ea-8947-d2f2c5e47a39.png)

If #925 (button mode) is merged, the text will be accurate. Otherwise the toolbar will open in float mode directly and there won't be any button to click...
 
## Checklist
- [ ] All querysets/queries filter by Team (if applicable)
- [x] Backend tests (if applicable)
- [ ] Cypress E2E tests (if applicable)

mariusandra committed:
per-user toggle to use new toolbar

mariusandra committed:
add migration

mariusandra committed:
fix test and remove setting

timgl commented on posthog/models/user.py:
```diff
@@ -78,6 +78,9 @@ class User(AbstractUser):
     anonymize_data: models.BooleanField = models.BooleanField(
         default=False, null=True, blank=True
     )
+    toolbar_mode: models.CharField = models.CharField(
```
Are you thinking we'll have different modes here? Might be worth adding those specifically [as choices](https://docs.djangoproject.com/en/3.0/ref/models/fields/#choices)

mariusandra replied:
I was thinking to reuse this field in the future and support values like "float" and "dock", to specify which would be the default mode for the "big floating button" that opens the toolbar.

So for now the 2 enums that this field uses are "default" and "toolbar". In the future I'd expand it to include "toolbar.float", "toolbar.dock" and who knows what else we can come up with.

Not sure if that makes sense or not. I can replace it with a bool for now and refactor/migrate it to a choice once we have a choice to give. The float/dock separation could even be in a new field then...

timgl replied:
Okay worth adding those two enums as Choices in that case, just so we don't end up with random stuff in there accidentally

mariusandra replied:
Choices added! They will for sure change when we release the toolbar, so I wouldn't put too much emphasis on having them perfect now.

mariusandra committed:
add choices to toolbar setting

mariusandra committed:
fix latest_migrations.manifest

mariusandra commented:
Hey, this is not yet merged... and this code is not found in any other PR either. I think this should be merged in whenever the larger toolbar PR is good enough to be somewhat useful.

timgl commented:
Sorry, not sure why I closed it! So this is waiting on #974

mariusandra commented:
Now this is waiting for #1033 . I'll rebase after that's in. Any changes to make to the text of the button (screenshot in the first post in this thread) to explain how experimental the toolbar is right now?

timgl commented:
> The toolbar gives you access to heatmaps, stats and allows you to create actions, without ever leaving your own website or app! Make sure you're using the snippet or posthog-js.

mariusandra committed:
update text

mariusandra committed:
add hint to restart the browser

mariusandra commented:
I rebased this PR and updated the text. Now it looks like this:

![image](https://user-images.githubusercontent.com/53387/85580832-5c3cc180-b63c-11ea-8f5a-8800bac518fd.png)

There's just one issue. If you used the toolbar already in your browser session (by having the toggle set for "new toolbar" and visiting your own site... or clicking "actions->inspect element" in posthog), then the toggle has no effect since the last `editorParams` are taken from `window.sessionStorage`. 

The fix then is to reload your browser or click "actions->inspect" again. To avoid needless support questions, I added this line to the "saved!" message:

![image](https://user-images.githubusercontent.com/53387/85581528-f3a21480-b63c-11ea-81fd-4e8e188168c2.png)

mariusandra committed:
fix test now that the setting is gone

mariusandra commented:
@timgl tagging you directly as this should be reviewed and merged as well

mariusandra commented:
@timgl This has been rebased!

timgl commented on frontend/src/scenes/setup/ToolbarSettings.js:
```diff
@@ -0,0 +1,44 @@
+                Try the new PostHog Toolbar?
+                {saved && (
+                    <span className="text-success" style={{ marginLeft: 10 }}>
+                        Preference saved. You might need to restart your browser session for the change to take effect.
```
I didn't understand "restart browser session" -- do you mean "restart my browser" or just refresh the page?

Either way when I went to 127.0.0.1:8000/demo it worked

mariusandra replied:
The first option - restarting the browser. With the current posthog-js, if you open the toolbar through "posthog -> actions -> inspect element on site", it'll jam the toolbar config as it was at that moment into `window.sessionStorage`. This includes information on which version of the toolbar to load.

Later if you visit your site, the data is taken from there... no matter what /decide replies. If you then use the toggle to use the new toolbar, it won't show up... as the data in sessionStorage will still force the old one to open.

This can be cleaned up, but that requires a new version of posthog-js. It's something I already started working on, but I want to get the tests working first (in progress).

timgl replied:
That makes sense, it'd be good to change the language to "restart your browser" which would be less confusing (for me at least!)

mariusandra replied:
Fixed!

timgl reviewed (approved):
One small nit but otherwise looks great! Tested both on an off state

mariusandra committed:
simplify language

timgl commented:
🎉
